File: java_version/bfs/Search_Graph_Node.java
package bfs;

import java.util.*;

public class Search_Graph_Node {
    public UndirectedGraphNode searchNode(ArrayList<UndirectedGraphNode> graph, Map<UndirectedGraphNode, Integer> values, UndirectedGraphNode node, int target) {
        UndirectedGraphNode result = null;
        if (graph == null){
            return result;
        }
        //use bfs to find shortest distance
        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.add(node);
        HashSet<UndirectedGraphNode> visited = new LinkedHashSet<>();
        visited.add(node);

        while(!queue.isEmpty()){
            UndirectedGraphNode new_node = queue.poll();
            if(values.get(new_node) == target){
                return new_node;
            }

            for(UndirectedGraphNode undirectedGraphNode: new_node.neighbors){
                if(visited.contains(undirectedGraphNode)){
                    continue;
                }
                if(values.get(undirectedGraphNode) == target){
                    return undirectedGraphNode;
                }

                visited.add(undirectedGraphNode);
                queue.add(undirectedGraphNode);
            }
        }
        return null;
    }
}

File: python_version/graph/search_graph_node.py
"""
Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []
"""


class Solution:
    """
    @param: graph: a list of Undirected graph node
    @param: values: a hash mapping, <UndirectedGraphNode, (int)value>
    @param: node: an Undirected graph node
    @param: target: An integer
    @return: a node
    """

    def searchNode(self, graph, values, node, target):
        # write your code here

        if values == []:
            return None
        if values[node] == target:
            return node

        # use bfs
        queue, visited = [node], []
        while queue != []:
            e = queue.pop(0)
            visited.append(e)
            if values[e] == target:
                return e

            for i in e.neighbors:
                if i not in visited:
                    queue.append(i)

        return None



